llama-cpp: pull upstream flake changes

happysalada opened the pull request:
## Description of changes

I've basically copied the upstream flake.
we had a discussion with @SomeoneSerge about removing the custom install step.
The idea was that using the normal cmake install step uses the upstream file. Whereas using a custom install step, we just loose upstream logic and we just do a worse job at it.

@elohmeier 
I've kept the naming of cudaSupport instead of upstream usecuda since it's the norm in nixpkgs, but I don't feel that strongly about it.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SomeoneSerge commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -45,94 +85,65 @@ effectiveStdenv.mkDerivation (finalAttrs: {
   '';
 
-  nativeBuildInputs = [ cmake ] ++ lib.optionals blasSupport [ pkg-config ] ++ lib.optionals cudaSupport [
+  nativeBuildInputs = [ cmake ninja pkg-config git ]
+    ++ optionals blasSupport [ pkg-config ]
```
Listing pkg-config twice

SomeoneSerge commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -45,94 +85,65 @@ effectiveStdenv.mkDerivation (finalAttrs: {
   cmakeFlags = [
-    "-DLLAMA_NATIVE=OFF"
-    "-DLLAMA_BUILD_SERVER=ON"
+    # reduces performance drastically on some architectures
+    (cmakeBool "LLAMA_NATIVE" false)
```
Iirc this corresponds to `-march=native`, so the comment should be "-march=native is non-deterministic, override with platform-specific flags instead"

SomeoneSerge commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -15,19 +15,59 @@
+      CoreVideo
+      CoreGraphics
+    ]
+    ++ optionals useMetalKit [ MetalKit ];
```
`metalSupport`, ofborg complains

SomeoneSerge commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -45,94 +85,65 @@ effectiveStdenv.mkDerivation (finalAttrs: {
+        # Build all targets supported by rocBLAS. When updating search for TARGET_LIST_ROCM
+        # in https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/CMakeLists.txt
+        # and select the line that matches the current nixpkgs version of rocBLAS.
+        # Should likely use `rocmPackages.clr.gpuTargets`.
+        "-DAMDGPU_TARGETS=gfx803;gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a:xnack+;gfx90a:xnack-;gfx940;gfx941;gfx942;gfx1010;gfx1012;gfx1030;gfx1100;gfx1101;gfx1102"
```
The comment is from the upstream flake, I guess it's ok to just "pull those changes", but consider following up on the suggestion from the comments in a separate commit

happysalada replied:
the xnack- present on some of these would make it difficult to abstract away
Happy to follow up with a separate PR when this has been resolved exactly and tested.

a-n-n-a-l-e-e commented:
see also https://github.com/NixOS/nixpkgs/pull/287554 -- think there is some overlap.

happysalada commented:
Thank you for letting me know, ill let them know.

mschwaig commented:
Thanks for asking me about this PR @happysalada, and about moving forward with this instead of mine.
In theory that's fine for me, but there's a few things I noticed about this PR in its current state that I want to mention:

If you are making such significant changes to a package, please give a short written explanation somewhere as part of the PR which explains what the reasons for and benefits of that change are. Saying discussions have been had with person X and Y helps a bit, but by itself it maybe makes the process feel even more opaque to readers of the PR, which I don't think should be acceptable. When I look at this PR I feel kind of left/locked out of the process.

As is, by getting rid of that custom installation step, 
* you're removing the `/lib` and `/include` folders. I can only assume that those are obsolete now or that the PR is not done yet, which you could clarify somewhere or mark it as a draft, and
* besides breaking the interface of the package, not renaming the content of the `/bin` folder as it was done previously looks like it will create issues if people put the package on their path, since it contains binaries with names like `lookup` and `simple`.

The other side of that issue of renaming binaries is that with whats currently on master it's actually really annoying to switch between something like ollama with llama.cpp from the upstream flake and nixpkgs, because of the different binary names. If you think about renaming all of the outputs, it would be great if you could look into addressing that issue as well.
I have been using this change to get around that: https://github.com/NixOS/nixpkgs/commit/9afa045adaf84417813d6d85cdf2d4e7c7a1ecf2

mschwaig commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -45,94 +85,64 @@ effectiveStdenv.mkDerivation (finalAttrs: {
+  # upstream plans on adding targets at the cmakelevel, remove those
+  # additional steps after that
+  postInstall = ''
+    mv $out/bin/main $out/bin/llama
```
this breaks `meta.mainProgram`

SomeoneSerge replied:
> really annoying to switch between something like ollama with llama.cpp from the upstream flake and nixpkgs, because of the different binary names

To be fair, it was the upstream's choice to use non-composable/generic names. 👍🏻 that the name should be consistent with whatever `meta.mainProgram` is kept

philiptaron replied:
How about this?

```suggestion
    mv $out/bin/main $out/bin/${finalAttrs.finalPackage.meta.mainProgram}
```

mschwaig commented on pkgs/by-name/ll/llama-cpp/package.nix:
`meta.broken` needs to be adapted to consider Vulkan. Right now the package is broken if only `vulkanSupport` is enabled. 

Right now `vulkanSupport` is also still broken on ARM Mac, and maybe we can get Intel Macs working by enabling `vulkanSupport` for them. See: https://github.com/NixOS/nixpkgs/pull/287554#issuecomment-1946505050

SomeoneSerge commented on pkgs/by-name/ll/llama-cpp/package.nix:
>     you're removing the /lib and /include folders

I didn't notice that, where? Libraries and headers are installed by CMake

mschwaig replied:
I built it locally and went by the output of `tree` to check. Embarrassingly I read that wrong. I will update my comment.

Actually the previous version commit only has `/bin`, and additionally `/lib` for `libggml_shared.so` and `libllama.so` with `.override{ static=false; }`.
This PR gets rid of the flag and has some extra stuff in `lib`and adds a `/include`.

```
result
├── bin
│   ├── baby-llama
│   ├── batched
│   ├── batched-bench
│   ├── beam-search
│   ├── benchmark
│   ├── convert-llama2c-to-ggml
│   ├── convert-lora-to-ggml.py
│   ├── convert.py
│   ├── embedding
│   ├── export-lora
│   ├── finetune
│   ├── gguf
│   ├── imatrix
│   ├── infill
│   ├── llama
│   ├── llama-bench
│   ├── llama-server
│   ├── llava-cli
│   ├── lookahead
│   ├── lookup
│   ├── parallel
│   ├── passkey
│   ├── perplexity
│   ├── quantize
│   ├── quantize-stats
│   ├── save-load-state
│   ├── simple
│   ├── speculative
│   ├── test-autorelease
│   ├── test-backend-ops
│   ├── test-grad0
│   ├── test-grammar-parser
│   ├── test-llama-grammar
│   ├── test-model-load-cancel
│   ├── test-quantize-fns
│   ├── test-quantize-perf
│   ├── test-rope
│   ├── test-sampling
│   ├── test-tokenizer-0-falcon
│   ├── test-tokenizer-0-llama
│   ├── test-tokenizer-1-bpe
│   ├── test-tokenizer-1-llama
│   ├── tokenize
│   └── train-text-from-scratch
├── include
│   ├── ggml-alloc.h
│   ├── ggml-backend.h
│   ├── ggml.h
│   └── llama.h
└── lib
    ├── cmake
    │   └── Llama
    │       ├── LlamaConfig.cmake
    │       └── LlamaConfigVersion.cmake
    ├── libggml_shared.so
    ├── libllama.so
    └── libllava_shared.so
```

SomeoneSerge replied:
Tbh I only checked the upstream flake yet. I'll have a look at both of the PRs during the week

SomeoneSerge commented on pkgs/by-name/ll/llama-cpp/package.nix:
> The other side of that issue of renaming binaries is that with whats currently on master it's actually really annoying to switch between something like ollama with llama.cpp from the upstream flake and nixpkgs, because of the different binary names

That's an interesting thought 👍🏻 . Since we in principle can contribute both to the nixpkgs and to the upstream nix expressions, we could expose all of the relevant metadata in the the `passthru` (or even in the outputs if needed) to accommodate smooth transition

SomeoneSerge commented on pkgs/by-name/ll/llama-cpp/package.nix:
> put the package on their path, since it contains binaries with names like lookup and simple

To ensure this doesn't bite us in the future, we could propose a PR upstream adding a cmake option that prefixes all of the installed binaries (the default prefix being the empty string)

mschwaig replied:
That's a good idea. 👍

philiptaron commented:
https://github.com/ggerganov/llama.cpp/pull/5311 has been merged and Vulkan works on macOS now.

philiptaron reviewed (changes requested):
Please integrate the changes from https://github.com/ggerganov/llama.cpp/pull/5311, then we can test on x86_64-linux, x86_64-darwin, and aarch64-darwin, ensure it works, then check it in.

https://github.com/NixOS/nixpkgs/pull/289513/files#r1493855899 also needs to be addressed in the nixpkgs context.

SomeoneSerge commented on pkgs/by-name/ll/llama-cpp/package.nix:
I'm maybe not following, but why integrating the Vulkan PR should be a blocker for merging this? I wouldn't mind if this was a separate PR, depending on whether it's easier to pull this together with or without the Vulkan enhancement.

@mschwaig's concerns need to be resolved indeed. Btw, @happysalada do you still have time for this?

philiptaron commented:
My goal is that macOS Nix users have the ability to get the accelerated Vulkan build. As @mschwaig [notes](https://github.com/NixOS/nixpkgs/pull/289513#discussion_r1493842801), with the current revision selected, vulkan needs a restriction that gates it to Linux only. Since this PR doesn't set `meta.broken` on Darwin in the first place if Vulkan support is specified, it only needs to be moved to a newer rev of llama.cpp in order to implement this feedback, I believe.

Super simple stuff, I think and hope.

happysalada commented:
Im happy to make the update at the same time, should have time for this saturday morning. Recently day job is crazy during the week.

happysalada commented:
Alright I've updated to latest 2252, I've rephrased the PR text, and I've added vulcan in the the non broken logic.

If anyone has any other comments on this PR please let me know.


Separately I have a question for @SomeoneSerge .
We had a discussion about ollama, that is doing a custom install step hardcoding a path for llama-cpp. You rightfully suggested that they should use pkg-config to locate the .so files. In order for ollama to be able to use package config, llama-cpp has to define a .pc file, correct ?
(you'll have to excuse my lack of knowledge in the c++/c build tooling system).

philiptaron commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -15,124 +15,134 @@
   effectiveStdenv = if cudaSupport then cudaPackages.backendStdenv else stdenv;
+  inherit (lib) cmakeBool cmakeFeature optionals;
+
+   # apple_sdk is supposed to choose sane defaults, no need to handle isAarch64
+  # separately
```
Can delete comment; this comment was composed with respect to an old version of flake.nix in the llama.cpp repository.

philiptaron commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -15,124 +15,134 @@
+    cuda_cccl.dev # <nv/target>
+
+    # A temporary hack for reducing the closure size, remove once cudaPackages
+    # have stopped using lndir: https://github.com/NixOS/nixpkgs/issues/271792
```
It's on the [Roadmap 🔮](https://github.com/NixOS/nixpkgs/issues/271792#event-de0a2e54-18e7-508c-84de-8f1c4528ea34)!

philiptaron commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -15,124 +15,134 @@
 effectiveStdenv.mkDerivation (finalAttrs: {
   pname = "llama-cpp";
-  version = "2249";
+  version = "2252";
```
Wow, we're already at 2275 just over the weekend 👀 . No need to adjust, but dang, this is why flakes enable Nix support in fast-moving repositories!

philiptaron commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -15,124 +15,134 @@
+    hash = "sha256-Gu7E6yWkxIjqENzSE+UtuQmUKEPgC/5WzZCekduB36E=";
   };
 
   postPatch = ''
     substituteInPlace ./ggml-metal.m \
       --replace '[bundle pathForResource:@"ggml-metal" ofType:@"metal"];' "@\"$out/bin/ggml-metal.metal\";"
   '';
```
Unrelated to your change: this is such an odd line and I wonder why the Nix build needs it. I'm no Metal-head, though.

philiptaron commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -15,124 +15,134 @@
-  ] ++ lib.optionals blasSupport [
-    openblas
-  ];
+  buildInputs = optionals effectiveStdenv.isDarwin darwinBuildInputs
+    ++ optionals cudaSupport cudaBuildInputs
+    ++ optionals mpiSupport mpi
+    ++ optionals openclSupport [ clblast ]
+    ++ optionals rocmSupport rocmBuildInputs
+    ++ optionals vulkanSupport vulkanBuildInputs;
```
Gorgeous.

philiptaron commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -15,124 +15,134 @@
   cmakeFlags = [
-    "-DLLAMA_NATIVE=OFF"
-    "-DLLAMA_BUILD_SERVER=ON"
+    # -march=native is non-deterministic override with platform-specific flags instead
```
```suggestion
    # -march=native is non-deterministic; override with platform-specific flags if needed
```

philiptaron commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -146,7 +156,7 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     mainProgram = "llama-cpp-main";
     maintainers = with maintainers; [ dit7ya elohmeier ];
-    broken = (effectiveStdenv.isDarwin && effectiveStdenv.isx86_64) || lib.count lib.id [openclSupport blasSupport rocmSupport cudaSupport] == 0;
+    broken = (effectiveStdenv.isDarwin && effectiveStdenv.isx86_64) || lib.count lib.id [openclSupport blasSupport rocmSupport cudaSupport vulkanSupport] == 0;
```
This isn't broken on x86_64-darwin anymore 🎉 
```suggestion
    broken = (effectiveStdenv.isDarwin && effectiveStdenv.isx86_64) || lib.count lib.id [openclSupport blasSupport rocmSupport cudaSupport] == 0;
```

philiptaron replied:
Wait, this check is just off. Upstream's version is `broken = (useMetalKit && !effectiveStdenv.isDarwin);` and that's correct.

philiptaron commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -146,7 +156,7 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     license = licenses.mit;
     mainProgram = "llama-cpp-main";
```
Upstream picks `llama` as the main program.
```suggestion
    mainProgram = "llama";
```

philiptaron commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -146,7 +156,7 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     license = licenses.mit;
     mainProgram = "llama-cpp-main";
     maintainers = with maintainers; [ dit7ya elohmeier ];
```
You can add me (and @SomeoneSerge if he accepts 🙇🏻 ).

```suggestion
    maintainers = with maintainers; [ dit7ya elohmeier philiptaron ];
```

philiptaron commented on pkgs/by-name/ll/llama-cpp/package.nix:
```diff
@@ -146,7 +156,7 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     license = licenses.mit;
```
The description also wants to be updated from the upstream: Inference of LLaMA model in pure C/C++

I like the `pnameSuffix` and `descriptionSuffix` things that upstream does as well, which I note you chose not to port over.

happysalada replied:
I was a little lazy, and wasn't 100% sure it was going to be accepted. I wanted to make the smallest change possible. How about we do that in a follow up PR ?

philiptaron replied:
Yeah, I have no strong feelings about it being in this PR.

philiptaron reviewed (changes requested):
Requesting changes for the `meta.mainProgram` thing and the `meta.broken` and `meta.badPlatforms` keys; a bunch of super small other things are also noted.

happysalada committed:
llama-cpp: 2249 -> 2294; bring upstream flake

happysalada commented:
Alright I went through the comments and fixed what was easy/simple to fix and brought the update to 2294.
Let me know if anything else.

philiptaron reviewed (approved):
Result of `nixpkgs-review pr 289513` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>llama-cpp</li>
  </ul>
</details>

philiptaron reviewed (approved):
Thanks @GrahamcOfBorg, I'm still saying yes.

philiptaron commented:
> @SomeoneSerge @philiptaron if any of you has time / availability to add a .pc file to llama-cpp ill be happy to try to propose using pkg-config to ollama . currently ollama is broken on nixpkgs because they want the lib files at specific places. Changing to pkg-config would fix that.

Agree, but I am a complete novice when it comes to `pkg-config`; the task is better suited for those who know how to use that means of ascent. I've got commit rights over in [ggerganov/llama.cpp](https://github.com/ggerganov/llama.cpp), so if you (or anyone here!) uncorks a patch, I'll see what I can do to get it merged.
